### PR TITLE
MAGECLOUD-3054: Implement Database Locks using Zookeeper and flock for Reliable HA Support

### PR DESCRIPTION
**Pull request by BaDos**
### Description
- Added patches that add support of Zookeeper and file locks for Magento 2.2.5-2.3.1
- Added autoconfiguration locks

### Fixed Issues (if relevant)
https://magento2.atlassian.net/browse/MAGECLOUD-3054

### Manual testing scenarios
https://jira.corp.magento.com/browse/MAGECLOUD-102

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\MagentoCloud\Process\Deploy\InstallUpdate\ConfigUpdate\Lock;
+
+use Magento\MagentoCloud\Config\Environment;
+
+/**
+ * Returns search configuration.
```
Comment (oshmyheliuk) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php]: Description isn't correct

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * @param Environment $environment
+     */
+    public function __construct(Environment $environment)
+    {
+        $this->environment = $environment;
+    }
+
+    /**
+     * Returns search configuration.
```
Comment (oshmyheliuk) [inline src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php]: Description isn't correct

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php (hunk 1 above)
Comment (BaDos) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php]: Fixed.

File: src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php (hunk 2 above)
Comment (BaDos) [reply src/Process/Deploy/InstallUpdate/ConfigUpdate/Lock/Config.php]: Fixed.

**Review by oshmyheliuk — APPROVED** [review]

**Review by NadiyaS — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request merged**